scipy/optimize/_gensa.py

```diff
@@ -0,0 +1,669 @@
+            return True
+        if self._nbfuncall >= self.maxfuncall:
+            self._message = ["Number of function call reached"]
+            return True
+
+    def _stop_search(self):
+        """Record time stamp when stop searching
+        """
+        if self.pure_sa:
+            # In case of pure sa approach, doing ad LS at the end
+            temp = np.array(self._xmin)
+            etemp = self._ls_energy(temp)
+            temp = self._xbuffer
+            self._xmin = np.array(temp)
+            self._emin = np.array(etemp)
+        self._endtime = time.time()
+
+    def _coordin(self):
+        """Random generation of new coordinates
+        """
+        self._x = self._random_state.random_sample(
+            self._x.size) * self._xrange + self._lower
+
+    def _energy(self, x):
+        """Calling objective function and adding elasticity if needed
+        """
+        delta_energy = 0
+        if self.has_constraint:
+            in_constraint = self.judge_constraint()
+            if not in_constraint:
+                self._etot = self.BIG_VALUE
+                return 0
+        if np.all(np.logical_and(
+                x >= self._lower, x <= self._upper)):
+            delta_energy = 0
+        else:
+            lcomp = x < self._lower
+            ucomp = x > self._upper
+            delta_energy_l = np.fabs(
+                x[lcomp] - self._lower[lcomp]) * self.KSPRING
+            delta_energy_u = np.fabs(
+                x[ucomp] - self._upper[ucomp]) * self.KSPRING
+            delta_energy = np.sum(delta_energy_l) + np.sum(
+                delta_energy_u)
+        self._etot = self.fun(x, *self.args)
+        self._nbfuncall += 1
+        self._etot = self._etot + delta_energy
+        if np.isinf(self._etot) or np.isnan(self._etot):
+            self._etot = self.BIG_VALUE
+
+    def _ls_energy(self, x):
+        """Performing a local search on the current location
+        """
+        self._xbuffer = np.array(x)
+        self._smooth_search()
+        self._x = np.array(self._xbuffer)
+        return self._fvalue
+
+    def _visita(self):
+        """Visiting distribution function
+        """
+        pi = np.arcsin(1.0) * 2.0
+        fator1 = np.exp(np.log(self._temperature) / (self.qv - 1.0))
+        fator2 = np.exp((4.0 - self.qv) * np.log(self.qv - 1.0))
+        fator3 = np.exp((2.0 - self.qv) * np.log(2.0) / (self.qv - 1.0))
+        fator4 = np.sqrt(pi) * fator1 * fator2 / (fator3 * (3.0 - self.qv))
+        fator5 = 1.0 / (self.qv - 1.0) - 0.5
+        d1 = 2.0 - fator5
+        fator6 = pi * (1.0 - fator5) / \
+            np.sin(pi * (1.0 - fator5)) / np.exp(gammaln(d1))
+        sigmax = np.exp(-(self.qv - 1.0) *
+                        np.log(fator6 / fator4) / (3.0 - self.qv))
+        x = sigmax * self._yygas()
+        y = self._yygas()
+        den = np.exp(
+            (self.qv - 1.0) * np.log((np.fabs(y))) / (3.0 - self.qv))
+        return x / den
+
```

Comment:
It seems like the visiting distribution could be refactored into it's own class.  I think GenSARunner has too much responsibility, and separating out the visiting distribution would clean things up and make it easier to unit test.